# UAS AMS
- Nama  : Gilbert Kenanya El Sarun 
- NIM   : A12.2020.06330 
- Kelas : A12.6502 
# Prediksi Pertandingan *EPL* dengan Metode Random Forest dan Rolling Averages

# Latar Belakang :
Tingkat minat masyarakat pada sepak bola saat ini sedang meningkat sangat tajam, hal ini dikarenakan baru selesainya pagelaran sepak bola terbesar yaitu Piala Dunia. Sehingga euforia masyarakat terhadap pertandingan sepak bola sangatlah tinggi. Tidak terlepas pula dari *EPL (English Premiere League)*, hal ini dikarenakan EPL merupakan sebuah pagelaran sepak bola yang bertabur klub – klub besar serta pemain bintang. Dalam menanggapi arus ini saya memilih menjadikan EPL menjadi object penelitian saya, karena EPL merupakan ajang sepak bola yang paling disorot setelah piala dunia serta memiliki catatan pertandingan yang seru serta hasil yang sulit ditebak oleh sebab itu akan menjadi topik yang menarik untuk memprediksi hasil dari EPL. 

# Metode yang Digunakan :
Dalam melakukan percobaan ini saya menggunakan 2 metode sekaligus yaitu Random Forest serta Rolling Averages

**Random Forest**
Sebuah metode algoritma yang dibangun dari kumpulan pohon keputusan (*Decision Tree*) yang dilatih dengan metode bagging yang mana kombinasi model pembelajaran meningkatkan hasil keseluruhan.

**Rolling Averages**
Kalkulasi untuk menganalisa point data dengan membuat serangkaian rata rata subset berbeda dari data lengkap.

# Pemodelan
Pemodelannya menggunakan sistem split dimana data akan dibagi menjadi 2 bagian yaitu data training dan juga Testing

Training (Data sebelum tahun 2022)
Testing  (Data sesudah tahun 2022)

# Proses Pengerjaan 
## Membaca data
Data yang digunakan adalah data pertandingan *EPL (English Premier League)* Musim ke 2020 dan 2021

In [1]:
import pandas as pd
pertandingan= pd.read_csv("matches_data.csv", index_col=0)
print ("Banyak Data : ", pertandingan.shape[0])
print ("Total Kolom : ", pertandingan.shape[1])
print ("Ambil 10 data teratas")
pertandingan.head(10)


Banyak Data :  1389
Total Kolom :  27
Ambil 10 data teratas


date   time            comp         round  day venue result   gf  \
1   2021-08-15  16:30  Premier League   Matchweek 1  Sun  Away      L  0.0   
2   2021-08-21  15:00  Premier League   Matchweek 2  Sat  Home      W  5.0   
3   2021-08-28  12:30  Premier League   Matchweek 3  Sat  Home      W  5.0   
4   2021-09-11  15:00  Premier League   Matchweek 4  Sat  Away      W  1.0   
6   2021-09-18  15:00  Premier League   Matchweek 5  Sat  Home      D  0.0   
8   2021-09-25  12:30  Premier League   Matchweek 6  Sat  Away      W  1.0   
10  2021-10-03  16:30  Premier League   Matchweek 7  Sun  Away      D  2.0   
11  2021-10-16  15:00  Premier League   Matchweek 8  Sat  Home      W  2.0   
13  2021-10-23  17:30  Premier League   Matchweek 9  Sat  Away      W  4.0   
15  2021-10-30  15:00  Premier League  Matchweek 10  Sat  Home      L  0.0   

     ga        opponent  ...  match report  notes    sh   sot  dist   fk   pk  \
1   1.0       Tottenham  ...  Match Report    NaN  18.0   4.0  16.9  1.0  0.0   
2   0.0    Norwich City  ...  Match Report    NaN  16.0   4.0  17.3  1.0  0.0   
3   0.0         Arsenal  ...  Match Report    NaN  25.0  10.0  14.3  0.0  0.0   
4   0.0  Leicester City  ...  Match Report    NaN  25.0   8.0  14.0  0.0  0.0   
6   0.0     Southampton  ...  Match Report    NaN  16.0   1.0  15.7  1.0  0.0   
8   0.0         Chelsea  ...  Match Report    NaN  15.0   3.0  16.5  0.0  0.0   
10  2.0       Liverpool  ...  Match Report    NaN  12.0   3.0  16.8  1.0  0.0   
11  0.0         Burnley  ...  Match Report    NaN  16.0   6.0  14.6  0.0  0.0   
13  1.0        Brighton  ...  Match Report    NaN  23.0  13.0  15.7  1.0  0.0   
15  2.0  Crystal Palace  ...  Match Report    NaN  14.0   3.0  17.0  0.0  0.0   

   pkatt  season             team  
1    0.0    2022  Manchester City  
2    0.0    2022  Manchester City  
3    0.0    2022  Manchester City  
4    0.0    2022  Manchester City  
6    0.0    2022  Manchester City  
8    0.0    2022  Manchester City  
10   0.0    2022  Manchester City  
11   0.0    2022  Manchester City  
13   0.0    2022  Manchester City  
15   0.0    2022  Manchester City  

[10 rows x 27 columns]

In [5]:
print("Total Pertandingan dari Setiap team")
pertandingan["team"].value_counts()

Total Pertandingan dari Setiap team


West Ham United             72
Brighton and Hove Albion    72
Southampton                 72
Manchester United           72
Newcastle United            72
Leeds United                71
Arsenal                     71
Manchester City             71
Wolverhampton Wanderers     71
Crystal Palace              71
Tottenham Hotspur           71
Burnley                     71
Chelsea                     70
Aston Villa                 70
Leicester City              70
Everton                     70
Sheffield United            38
West Bromwich Albion        38
Liverpool                   38
Fulham                      38
Brentford                   34
Norwich City                33
Watford                     33
Name: team, dtype: int64

Dilihat disini ada beberapa team yang hanya mendapatkan pertandingan dibawah 40 kali, hal ini dikarenakan
mereka hanya mengikuti 1 musim saja diantara musim 2020 dan 2021

Sebagai Contoh Pertandingan dari Fulham

In [13]:
pertandingan[pertandingan["team"]=="Fulham"]

date   time            comp         round  day venue result   gf  \
0   2020-09-12  12:30  Premier League   Matchweek 1  Sat  Home      L  0.0   
2   2020-09-19  15:00  Premier League   Matchweek 2  Sat  Away      L  3.0   
4   2020-09-28  17:45  Premier League   Matchweek 3  Mon  Home      L  0.0   
6   2020-10-04  14:00  Premier League   Matchweek 4  Sun  Away      L  0.0   
7   2020-10-18  12:00  Premier League   Matchweek 5  Sun  Away      D  1.0   
8   2020-10-24  15:00  Premier League   Matchweek 6  Sat  Home      L  1.0   
9   2020-11-02  17:30  Premier League   Matchweek 7  Mon  Home      W  2.0   
10  2020-11-07  20:00  Premier League   Matchweek 8  Sat  Away      L  0.0   
11  2020-11-22  12:00  Premier League   Matchweek 9  Sun  Home      L  2.0   
12  2020-11-30  17:30  Premier League  Matchweek 10  Mon  Away      W  2.0   
13  2020-12-05  15:00  Premier League  Matchweek 11  Sat  Away      L  0.0   
14  2020-12-13  16:30  Premier League  Matchweek 12  Sun  Home      D  1.0   
15  2020-12-16  20:00  Premier League  Matchweek 13  Wed  Home      D  0.0   
16  2020-12-19  20:00  Premier League  Matchweek 14  Sat  Away      D  1.0   
17  2020-12-26  15:00  Premier League  Matchweek 15  Sat  Home      D  0.0   
19  2021-01-13  20:15  Premier League  Matchweek 16  Wed  Away      D  1.0   
20  2021-01-16  17:30  Premier League  Matchweek 19  Sat  Home      L  0.0   
21  2021-01-20  20:15  Premier League  Matchweek 18  Wed  Home      L  1.0   
23  2021-01-27  19:30  Premier League  Matchweek 20  Wed  Away      D  0.0   
24  2021-01-30  15:00  Premier League  Matchweek 21  Sat  Away      D  2.0   
25  2021-02-03  18:00  Premier League  Matchweek 22  Wed  Home      L  0.0   
26  2021-02-06  17:30  Premier League  Matchweek 23  Sat  Home      D  0.0   
27  2021-02-14  19:00  Premier League  Matchweek 24  Sun  Away      W  2.0   
28  2021-02-17  18:00  Premier League  Matchweek 17  Wed  Away      D  1.0   
29  2021-02-20  20:00  Premier League  Matchweek 25  Sat  Home      W  1.0   
30  2021-02-28  12:00  Premier League  Matchweek 26  Sun  Away      D  0.0   
31  2021-03-04  18:00  Premier League  Matchweek 33  Thu  Home      L  0.0   
32  2021-03-07  14:00  Premier League  Matchweek 27  Sun  Away      W  1.0   
33  2021-03-13  20:00  Premier League  Matchweek 28  Sat  Home      L  0.0   
34  2021-03-19  20:00  Premier League  Matchweek 29  Fri  Home      L  1.0   
35  2021-04-04  16:30  Premier League  Matchweek 30  Sun  Away      L  1.0   
36  2021-04-09  20:00  Premier League  Matchweek 31  Fri  Home      L  0.0   
37  2021-04-18  13:30  Premier League  Matchweek 32  Sun  Away      D  1.0   
38  2021-05-01  17:30  Premier League  Matchweek 34  Sat  Away      L  0.0   
39  2021-05-10  20:00  Premier League  Matchweek 35  Mon  Home      L  0.0   
40  2021-05-15  15:00  Premier League  Matchweek 36  Sat  Away      L  1.0   
41  2021-05-18  18:00  Premier League  Matchweek 37  Tue  Away      D  1.0   
42  2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      L  0.0   

     ga         opponent  ...  match report  notes    sh  sot  dist   fk   pk  \
0   3.0          Arsenal  ...  Match Report    NaN   5.0  2.0  26.0  0.0  0.0   
2   4.0     Leeds United  ...  Match Report    NaN  13.0  5.0  20.9  1.0  1.0   
4   3.0      Aston Villa  ...  Match Report    NaN  16.0  2.0  16.1  0.0  0.0   
6   1.0           Wolves  ...  Match Report    NaN  10.0  2.0  17.5  1.0  0.0   
7   1.0    Sheffield Utd  ...  Match Report    NaN  14.0  6.0  16.6  0.0  0.0   
8   2.0   Crystal Palace  ...  Match Report    NaN  16.0  3.0  18.2  0.0  0.0   
9   0.0        West Brom  ...  Match Report    NaN  13.0  4.0  17.8  0.0  0.0   
10  1.0         West Ham  ...  Match Report    NaN  12.0  2.0  19.7  1.0  0.0   
11  3.0          Everton  ...  Match Report    NaN  13.0  6.0  19.9  0.0  0.0   
12  1.0   Leicester City  ...  Match Report    NaN   9.0  4.0  20.4  1.0  1.0   
13  2.0  Manchester City  ...  Match Report    NaN   6.0  1.0  19.7  0.0  0.

Disini terlihat bahwa Fulham hanya memiliki data pertandingan pada musim 2020-2021 hal ini dikarenakan pada
musim selanjutnya Fulham terdegradasi dari EPL sehingga tidak bisa mengikuti pertandingan dimusim selanjutnya

# Membersihkan data kosong

In [18]:
print("Malihat jenis data masing - masing kolom")
pertandingan.dtypes

Malihat jenis data masing - masing kolom


date             object
time             object
comp             object
round            object
day              object
venue            object
result           object
gf              float64
ga              float64
opponent         object
xg              float64
xga             float64
poss            float64
attendance      float64
captain          object
formation        object
referee          object
match report     object
notes           float64
sh              float64
sot             float64
dist            float64
fk              float64
pk              float64
pkatt           float64
season            int64
team             object
dtype: object

In [21]:
pertandingan["date"]=pd.to_datetime(pertandingan["date"])
pertandingan

date   time            comp         round  day venue result   gf  \
1  2021-08-15  16:30  Premier League   Matchweek 1  Sun  Away      L  0.0   
2  2021-08-21  15:00  Premier League   Matchweek 2  Sat  Home      W  5.0   
3  2021-08-28  12:30  Premier League   Matchweek 3  Sat  Home      W  5.0   
4  2021-09-11  15:00  Premier League   Matchweek 4  Sat  Away      W  1.0   
6  2021-09-18  15:00  Premier League   Matchweek 5  Sat  Home      D  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
38 2021-05-02  19:15  Premier League  Matchweek 34  Sun  Away      L  0.0   
39 2021-05-08  15:00  Premier League  Matchweek 35  Sat  Home      L  0.0   
40 2021-05-16  19:00  Premier League  Matchweek 36  Sun  Away      W  1.0   
41 2021-05-19  18:00  Premier League  Matchweek 37  Wed  Away      L  0.0   
42 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  1.0   

     ga        opponent  ...  match report  notes    sh   sot  dist   fk   pk  \
1   1.0       Tottenham  ...  Match Report    NaN  18.0   4.0  16.9  1.0  0.0   
2   0.0    Norwich City  ...  Match Report    NaN  16.0   4.0  17.3  1.0  0.0   
3   0.0         Arsenal  ...  Match Report    NaN  25.0  10.0  14.3  0.0  0.0   
4   0.0  Leicester City  ...  Match Report    NaN  25.0   8.0  14.0  0.0  0.0   
6   0.0     Southampton  ...  Match Report    NaN  16.0   1.0  15.7  1.0  0.0   
..  ...             ...  ...           ...    ...   ...   ...   ...  ...  ...   
38  4.0       Tottenham  ...  Match Report    NaN   8.0   1.0  17.4  0.0  0.0   
39  2.0  Crystal Palace  ...  Match Report    NaN   7.0   0.0  11.4  1.0  0.0   
40  0.0         Everton  ...  Match Report    NaN  10.0   3.0  17.0  0.0  0.0   
41  1.0   Newcastle Utd  ...  Match Report    NaN  11.0   1.0  16.0  1.0  0.0   
42  0.0         Burnley  ...  Match Report    NaN  12.0   3.0  17.0  0.0  0.0   

   pkatt  season              team  
1    0.0    2022   Manchester City  
2    0.0    2022   Manchester City  
3    0.0    2022   Manchester City  
4    0.0    2022   Manchester City  
6    0.0    2022   Manchester City  
..   ...     ...               ...  
38   0.0    2021  Sheffield United  
39   0.0    2021  Sheffield United  
40   0.0    2021  Sheffield United  
41   0.0    2021  Sheffield United  
42   0.0    2021  Sheffield United  

[1389 rows x 27 columns]

In [22]:
pertandingan.dtypes

date            datetime64[ns]
time                    object
comp                    object
round                   object
day                     object
venue                   object
result                  object
gf                     float64
ga                     float64
opponent                object
xg                     float64
xga                    float64
poss                   float64
attendance             float64
captain                 object
formation               object
referee                 object
match report            object
notes                  float64
sh                     float64
sot                    float64
dist                   float64
fk                     float64
pk                     float64
pkatt                  float64
season                   int64
team                    object
dtype: object

Dari keterangan ini terlihat bahwa kolom date sudah terubah menjadi tipe data datetime yang mana nantinya
akan berguna dalam proses selanjutnya. karena pada machine learning nantinya tidak dapat mengolah tipe data
object

# Membuat Acuan
## Menambahkan Kolom yang berisi informasi berguna nantinya

Menambahkan kolom kode_lokasi yang berfungsi untuk mengubah jenis data dari kolom venue kedalam integer dengan pengkodean 0 untuk away, dan 1 untuk home

In [25]:
pertandingan["kode_lokasi"]=pertandingan["venue"].astype("category").cat.codes
pertandingan

date   time            comp         round  day venue result   gf  \
1  2021-08-15  16:30  Premier League   Matchweek 1  Sun  Away      L  0.0   
2  2021-08-21  15:00  Premier League   Matchweek 2  Sat  Home      W  5.0   
3  2021-08-28  12:30  Premier League   Matchweek 3  Sat  Home      W  5.0   
4  2021-09-11  15:00  Premier League   Matchweek 4  Sat  Away      W  1.0   
6  2021-09-18  15:00  Premier League   Matchweek 5  Sat  Home      D  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
38 2021-05-02  19:15  Premier League  Matchweek 34  Sun  Away      L  0.0   
39 2021-05-08  15:00  Premier League  Matchweek 35  Sat  Home      L  0.0   
40 2021-05-16  19:00  Premier League  Matchweek 36  Sun  Away      W  1.0   
41 2021-05-19  18:00  Premier League  Matchweek 37  Wed  Away      L  0.0   
42 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  1.0   

     ga        opponent  ...  notes    sh   sot  dist   fk   pk pkatt season  \
1   1.0       Tottenham  ...    NaN  18.0   4.0  16.9  1.0  0.0   0.0   2022   
2   0.0    Norwich City  ...    NaN  16.0   4.0  17.3  1.0  0.0   0.0   2022   
3   0.0         Arsenal  ...    NaN  25.0  10.0  14.3  0.0  0.0   0.0   2022   
4   0.0  Leicester City  ...    NaN  25.0   8.0  14.0  0.0  0.0   0.0   2022   
6   0.0     Southampton  ...    NaN  16.0   1.0  15.7  1.0  0.0   0.0   2022   
..  ...             ...  ...    ...   ...   ...   ...  ...  ...   ...    ...   
38  4.0       Tottenham  ...    NaN   8.0   1.0  17.4  0.0  0.0   0.0   2021   
39  2.0  Crystal Palace  ...    NaN   7.0   0.0  11.4  1.0  0.0   0.0   2021   
40  0.0         Everton  ...    NaN  10.0   3.0  17.0  0.0  0.0   0.0   2021   
41  1.0   Newcastle Utd  ...    NaN  11.0   1.0  16.0  1.0  0.0   0.0   2021   
42  0.0         Burnley  ...    NaN  12.0   3.0  17.0  0.0  0.0   0.0   2021   

                team  kode_lokasi  
1    Manchester City            0  
2    Manchester City            1  
3    Manchester City            1  
4    Manchester City            0  
6    Manchester City            1  
..               ...          ...  
38  Sheffield United            0  
39  Sheffield United            1  
40  Sheffield United            0  
41  Sheffield United            0  
42  Sheffield United            1  

[1389 rows x 28 columns]

Menambahkan kolom kode_team untuk mengubah data nama team menjadi pengkodean angka sehingga dapat diproses dalam machine learning nantinya

In [26]:
pertandingan["kode_team"]=pertandingan["opponent"].astype("category").cat.codes
pertandingan

date   time            comp         round  day venue result   gf  \
1  2021-08-15  16:30  Premier League   Matchweek 1  Sun  Away      L  0.0   
2  2021-08-21  15:00  Premier League   Matchweek 2  Sat  Home      W  5.0   
3  2021-08-28  12:30  Premier League   Matchweek 3  Sat  Home      W  5.0   
4  2021-09-11  15:00  Premier League   Matchweek 4  Sat  Away      W  1.0   
6  2021-09-18  15:00  Premier League   Matchweek 5  Sat  Home      D  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
38 2021-05-02  19:15  Premier League  Matchweek 34  Sun  Away      L  0.0   
39 2021-05-08  15:00  Premier League  Matchweek 35  Sat  Home      L  0.0   
40 2021-05-16  19:00  Premier League  Matchweek 36  Sun  Away      W  1.0   
41 2021-05-19  18:00  Premier League  Matchweek 37  Wed  Away      L  0.0   
42 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  1.0   

     ga        opponent  ...    sh   sot  dist   fk   pk pkatt season  \
1   1.0       Tottenham  ...  18.0   4.0  16.9  1.0  0.0   0.0   2022   
2   0.0    Norwich City  ...  16.0   4.0  17.3  1.0  0.0   0.0   2022   
3   0.0         Arsenal  ...  25.0  10.0  14.3  0.0  0.0   0.0   2022   
4   0.0  Leicester City  ...  25.0   8.0  14.0  0.0  0.0   0.0   2022   
6   0.0     Southampton  ...  16.0   1.0  15.7  1.0  0.0   0.0   2022   
..  ...             ...  ...   ...   ...   ...  ...  ...   ...    ...   
38  4.0       Tottenham  ...   8.0   1.0  17.4  0.0  0.0   0.0   2021   
39  2.0  Crystal Palace  ...   7.0   0.0  11.4  1.0  0.0   0.0   2021   
40  0.0         Everton  ...  10.0   3.0  17.0  0.0  0.0   0.0   2021   
41  1.0   Newcastle Utd  ...  11.0   1.0  16.0  1.0  0.0   0.0   2021   
42  0.0         Burnley  ...  12.0   3.0  17.0  0.0  0.0   0.0   2021   

                team  kode_lokasi  kode_team  
1    Manchester City            0         18  
2    Manchester City            1         15  
3    Manchester City            1          0  
4    Manchester City            0         10  
6    Manchester City            1         17  
..               ...          ...        ...  
38  Sheffield United            0         18  
39  Sheffield United            1          6  
40  Sheffield United            0          7  
41  Sheffield United            0         14  
42  Sheffield United            1          4  

[1389 rows x 29 columns]

Menambahkan kolom jam dari kolom time menjadi angka, untuk menjadi salah satu acuan nantinya apabila misal suatu team bermain bagus di jam tertentu

In [35]:
pertandingan["jam"]= pertandingan["time"].str.replace(":.+", "", regex=True).astype("int")
pertandingan

date   time            comp         round  day venue result   gf  \
1  2021-08-15  16:30  Premier League   Matchweek 1  Sun  Away      L  0.0   
2  2021-08-21  15:00  Premier League   Matchweek 2  Sat  Home      W  5.0   
3  2021-08-28  12:30  Premier League   Matchweek 3  Sat  Home      W  5.0   
4  2021-09-11  15:00  Premier League   Matchweek 4  Sat  Away      W  1.0   
6  2021-09-18  15:00  Premier League   Matchweek 5  Sat  Home      D  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
38 2021-05-02  19:15  Premier League  Matchweek 34  Sun  Away      L  0.0   
39 2021-05-08  15:00  Premier League  Matchweek 35  Sat  Home      L  0.0   
40 2021-05-16  19:00  Premier League  Matchweek 36  Sun  Away      W  1.0   
41 2021-05-19  18:00  Premier League  Matchweek 37  Wed  Away      L  0.0   
42 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  1.0   

     ga        opponent  ...   sot  dist   fk   pk pkatt season  \
1   1.0       Tottenham  ...   4.0  16.9  1.0  0.0   0.0   2022   
2   0.0    Norwich City  ...   4.0  17.3  1.0  0.0   0.0   2022   
3   0.0         Arsenal  ...  10.0  14.3  0.0  0.0   0.0   2022   
4   0.0  Leicester City  ...   8.0  14.0  0.0  0.0   0.0   2022   
6   0.0     Southampton  ...   1.0  15.7  1.0  0.0   0.0   2022   
..  ...             ...  ...   ...   ...  ...  ...   ...    ...   
38  4.0       Tottenham  ...   1.0  17.4  0.0  0.0   0.0   2021   
39  2.0  Crystal Palace  ...   0.0  11.4  1.0  0.0   0.0   2021   
40  0.0         Everton  ...   3.0  17.0  0.0  0.0   0.0   2021   
41  1.0   Newcastle Utd  ...   1.0  16.0  1.0  0.0   0.0   2021   
42  0.0         Burnley  ...   3.0  17.0  0.0  0.0   0.0   2021   

                team kode_lokasi  kode_team  jam  
1    Manchester City           0         18   16  
2    Manchester City           1         15   15  
3    Manchester City           1          0   12  
4    Manchester City           0         10   15  
6    Manchester City           1         17   15  
..               ...         ...        ...  ...  
38  Sheffield United           0         18   19  
39  Sheffield United           1          6   15  
40  Sheffield United           0          7   19  
41  Sheffield United           0         14   18  
42  Sheffield United           1          4   16  

[1389 rows x 30 columns]

Menambahkan kolom hari untuk mengubah kolom day menjadi pengkodean angka

In [39]:
pertandingan["hari"]=pertandingan["date"].dt.dayofweek
pertandingan

date   time            comp         round  day venue result   gf  \
1  2021-08-15  16:30  Premier League   Matchweek 1  Sun  Away      L  0.0   
2  2021-08-21  15:00  Premier League   Matchweek 2  Sat  Home      W  5.0   
3  2021-08-28  12:30  Premier League   Matchweek 3  Sat  Home      W  5.0   
4  2021-09-11  15:00  Premier League   Matchweek 4  Sat  Away      W  1.0   
6  2021-09-18  15:00  Premier League   Matchweek 5  Sat  Home      D  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
38 2021-05-02  19:15  Premier League  Matchweek 34  Sun  Away      L  0.0   
39 2021-05-08  15:00  Premier League  Matchweek 35  Sat  Home      L  0.0   
40 2021-05-16  19:00  Premier League  Matchweek 36  Sun  Away      W  1.0   
41 2021-05-19  18:00  Premier League  Matchweek 37  Wed  Away      L  0.0   
42 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  1.0   

     ga        opponent  ...  dist   fk   pk  pkatt season              team  \
1   1.0       Tottenham  ...  16.9  1.0  0.0    0.0   2022   Manchester City   
2   0.0    Norwich City  ...  17.3  1.0  0.0    0.0   2022   Manchester City   
3   0.0         Arsenal  ...  14.3  0.0  0.0    0.0   2022   Manchester City   
4   0.0  Leicester City  ...  14.0  0.0  0.0    0.0   2022   Manchester City   
6   0.0     Southampton  ...  15.7  1.0  0.0    0.0   2022   Manchester City   
..  ...             ...  ...   ...  ...  ...    ...    ...               ...   
38  4.0       Tottenham  ...  17.4  0.0  0.0    0.0   2021  Sheffield United   
39  2.0  Crystal Palace  ...  11.4  1.0  0.0    0.0   2021  Sheffield United   
40  0.0         Everton  ...  17.0  0.0  0.0    0.0   2021  Sheffield United   
41  1.0   Newcastle Utd  ...  16.0  1.0  0.0    0.0   2021  Sheffield United   
42  0.0         Burnley  ...  17.0  0.0  0.0    0.0   2021  Sheffield United   

   kode_lokasi kode_team  jam  hari  
1            0        18   16     6  
2            1        15   15     5  
3            1         0   12     5  
4            0        10   15     5  
6            1        17   15     5  
..         ...       ...  ...   ...  
38           0        18   19     6  
39           1         6   15     5  
40           0         7   19     6  
41           0        14   18     2  
42           1         4   16     6  

[1389 rows x 31 columns]

Menambahkan Kolom target yang berfungsi menjadi acuan utama, di mana dalam kolom ini berisikan hasil pertandingan yang ada, apabila menang akan berisi 1 jika kalah atau seimbang akan berisi 0 (Karena kita akan memprediksi team yang menang nantinya)

In [40]:
pertandingan["target"]=(pertandingan["result"]=="W").astype("int")
pertandingan

date   time            comp         round  day venue result   gf  \
1  2021-08-15  16:30  Premier League   Matchweek 1  Sun  Away      L  0.0   
2  2021-08-21  15:00  Premier League   Matchweek 2  Sat  Home      W  5.0   
3  2021-08-28  12:30  Premier League   Matchweek 3  Sat  Home      W  5.0   
4  2021-09-11  15:00  Premier League   Matchweek 4  Sat  Away      W  1.0   
6  2021-09-18  15:00  Premier League   Matchweek 5  Sat  Home      D  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
38 2021-05-02  19:15  Premier League  Matchweek 34  Sun  Away      L  0.0   
39 2021-05-08  15:00  Premier League  Matchweek 35  Sat  Home      L  0.0   
40 2021-05-16  19:00  Premier League  Matchweek 36  Sun  Away      W  1.0   
41 2021-05-19  18:00  Premier League  Matchweek 37  Wed  Away      L  0.0   
42 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  1.0   

     ga        opponent  ...   fk   pk  pkatt  season              team  \
1   1.0       Tottenham  ...  1.0  0.0    0.0    2022   Manchester City   
2   0.0    Norwich City  ...  1.0  0.0    0.0    2022   Manchester City   
3   0.0         Arsenal  ...  0.0  0.0    0.0    2022   Manchester City   
4   0.0  Leicester City  ...  0.0  0.0    0.0    2022   Manchester City   
6   0.0     Southampton  ...  1.0  0.0    0.0    2022   Manchester City   
..  ...             ...  ...  ...  ...    ...     ...               ...   
38  4.0       Tottenham  ...  0.0  0.0    0.0    2021  Sheffield United   
39  2.0  Crystal Palace  ...  1.0  0.0    0.0    2021  Sheffield United   
40  0.0         Everton  ...  0.0  0.0    0.0    2021  Sheffield United   
41  1.0   Newcastle Utd  ...  1.0  0.0    0.0    2021  Sheffield United   
42  0.0         Burnley  ...  0.0  0.0    0.0    2021  Sheffield United   

   kode_lokasi kode_team jam  hari  target  
1            0        18  16     6       0  
2            1        15  15     5       1  
3            1         0  12     5       1  
4            0        10  15     5       1  
6            1        17  15     5       0  
..         ...       ...  ..   ...     ...  
38           0        18  19     6       0  
39           1         6  15     5       0  
40           0         7  19     6       1  
41           0        14  18     2       0  
42           1         4  16     6       1  

[1389 rows x 32 columns]

# Melakukan Training data (Random Forest)

In [42]:
!pip install scikit-learn

In [45]:
from sklearn.ensemble import RandomForestClassifier

## Menginisiasi Fungsi

In [48]:
rf= RandomForestClassifier(n_estimators=50, min_samples_split=10, random_state=1)

Membagi data ke dalam 2 (train, test), serta menginisiasi acuan yang digunakan

In [50]:
train=pertandingan[pertandingan["date"] < '2022-01-01']
test=pertandingan[pertandingan["date"] > '2022-01-01']
acuan=["kode_lokasi", "kode_team", "jam", "hari"]

membuat fungsi untuk melatih

In [51]:
rf.fit(train[acuan], train["target"])

RandomForestClassifier(min_samples_split=10, n_estimators=50, random_state=1)

In [52]:
prediksi = rf.predict(test[acuan])

# Melihat Akurasi dari model

In [54]:
from sklearn.metrics import accuracy_score
acc = accuracy_score(test["target"], prediksi)
acc

0.6123188405797102

In [55]:
combined = pd.DataFrame(dict(actual=test["target"], prediction=prediksi))

In [56]:
pd.crosstab(index=combined["actual"], columns=combined["prediction"])

prediction    0   1
actual             
0           141  31
1            76  28

Dari olahan ini terlihat bahwa :

- Akurasi data kita sebesar = 0.6123188405797102
- Berdasarkan data tersebut terlihat pada saat kita memprediksi tim yang kalah ataupun seimbang kita memperoleh hasil yang bagus yaitu 141 benar dan 76 salah
- Akan tetapi untuk memprediksi tim yang menang kita memperoleh hasil yang kurang memuaskan yaitu 31 salah dan 28 benar

Sehingga kita harus merevisi akurasi kita menggunakan metode lainnya

In [57]:
from sklearn.metrics import precision_score

In [58]:
precision_score(test["target"], prediksi)

0.4745762711864407

Dari penghitungan akurasi selanjutnya terlihat bahwa akurasi kita hanya 0.4745762711864407 sehingga sesuai dengan data, bahwa machine kita kurang akurat dalam memprediksi

# Meningkatkan akurasi dengan Rolling Averages

### Membagi data set menjadi beberapa data set tergantung team

In [59]:
group_pertandingan= pertandingan.groupby("team")

In [62]:
print("Melihat Permainan Manchester City")
group=group_pertandingan.get_group("Manchester City")
group

Melihat Permainan Manchester City


date   time            comp         round  day venue result   gf  \
1  2021-08-15  16:30  Premier League   Matchweek 1  Sun  Away      L  0.0   
2  2021-08-21  15:00  Premier League   Matchweek 2  Sat  Home      W  5.0   
3  2021-08-28  12:30  Premier League   Matchweek 3  Sat  Home      W  5.0   
4  2021-09-11  15:00  Premier League   Matchweek 4  Sat  Away      W  1.0   
6  2021-09-18  15:00  Premier League   Matchweek 5  Sat  Home      D  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
54 2021-05-01  12:30  Premier League  Matchweek 34  Sat  Away      W  2.0   
56 2021-05-08  17:30  Premier League  Matchweek 35  Sat  Home      L  1.0   
57 2021-05-14  20:00  Premier League  Matchweek 36  Fri  Away      W  4.0   
58 2021-05-18  19:00  Premier League  Matchweek 37  Tue  Away      L  2.0   
59 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  5.0   

     ga        opponent  ...   fk   pk  pkatt  season             team  \
1   1.0       Tottenham  ...  1.0  0.0    0.0    2022  Manchester City   
2   0.0    Norwich City  ...  1.0  0.0    0.0    2022  Manchester City   
3   0.0         Arsenal  ...  0.0  0.0    0.0    2022  Manchester City   
4   0.0  Leicester City  ...  0.0  0.0    0.0    2022  Manchester City   
6   0.0     Southampton  ...  1.0  0.0    0.0    2022  Manchester City   
..  ...             ...  ...  ...  ...    ...     ...              ...   
54  0.0  Crystal Palace  ...  1.0  0.0    0.0    2021  Manchester City   
56  2.0         Chelsea  ...  0.0  0.0    1.0    2021  Manchester City   
57  3.0   Newcastle Utd  ...  1.0  0.0    0.0    2021  Manchester City   
58  3.0        Brighton  ...  1.0  0.0    0.0    2021  Manchester City   
59  0.0         Everton  ...  0.0  0.0    0.0    2021  Manchester City   

   kode_lokasi kode_team jam  hari  target  
1            0        18  16     6       0  
2            1        15  15     5       1  
3            1         0  12     5       1  
4            0        10  15     5       1  
6            1        17  15     5       0  
..         ...       ...  ..   ...     ...  
54           0         6  12     5       1  
56           1         5  17     5       0  
57           0        14  20     4       1  
58           0         3  19     1       0  
59           1         7  16     6       1  

[71 rows x 32 columns]

Dalam penggunaan metode Rolling Averages data akan menggunakan hasil dari pertandingan sebelumnya apakah tim tersebut menang atau kalah dan apa yang terjadi dalam pertandingan sebelumnya(3 Pertandingan) yang nantinya akan diolah ke dalam algoritma

In [72]:
def rolling_averages(group, cols, new_col):
    group=group.sort_values("date")
    rolling_stats=group[cols].rolling(3, closed='left').mean()
    group[new_cols]= rolling_stats
    group=group.dropna(subset=new_cols)
    return group

In [73]:
cols=["gf", "ga", "sh", "sot", "dist", "fk", "pk", "pkatt"]
new_cols=[f"{c}_rolling" for c in cols]
new_cols

['gf_rolling',
 'ga_rolling',
 'sh_rolling',
 'sot_rolling',
 'dist_rolling',
 'fk_rolling',
 'pk_rolling',
 'pkatt_rolling']

In [74]:
rolling_averages(group, cols, new_cols)

date   time            comp         round  day venue result   gf  \
5  2020-10-17  17:30  Premier League   Matchweek 5  Sat  Home      W  1.0   
7  2020-10-24  12:30  Premier League   Matchweek 6  Sat  Away      D  1.0   
9  2020-10-31  12:30  Premier League   Matchweek 7  Sat  Away      W  1.0   
11 2020-11-08  16:30  Premier League   Matchweek 8  Sun  Home      D  1.0   
12 2020-11-21  17:30  Premier League   Matchweek 9  Sat  Away      L  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
42 2022-03-14  20:00  Premier League  Matchweek 29  Mon  Away      D  0.0   
44 2022-04-02  15:00  Premier League  Matchweek 31  Sat  Away      W  2.0   
46 2022-04-10  16:30  Premier League  Matchweek 32  Sun  Home      D  2.0   
49 2022-04-20  20:00  Premier League  Matchweek 30  Wed  Home      W  3.0   
50 2022-04-23  15:00  Premier League  Matchweek 34  Sat  Home      W  5.0   

     ga        opponent  ...  hari  target  gf_rolling  ga_rolling sh_rolling  \
5   0.0         Arsenal  ...     5       1    2.000000    2.333333  17.333333   
7   1.0        West Ham  ...     5       0    1.333333    2.000000  17.333333   
9   0.0   Sheffield Utd  ...     5       1    1.000000    0.666667  16.666667   
11  1.0       Liverpool  ...     6       0    1.000000    0.333333  14.333333   
12  2.0       Tottenham  ...     5       0    1.000000    0.666667  12.000000   
..  ...             ...  ...   ...     ...         ...         ...        ...   
42  0.0  Crystal Palace  ...     0       0    2.333333    1.333333  19.000000   
44  0.0         Burnley  ...     5       1    1.666667    0.333333  18.333333   
46  2.0       Liverpool  ...     6       0    2.000000    0.333333  20.000000   
49  0.0        Brighton  ...     2       1    1.333333    0.666667  15.666667   
50  1.0         Watford  ...     5       1    2.333333    0.666667  15.333333   

   sot_rolling dist_rolling fk_rolling  pk_rolling  pkatt_rolling  
5     4.666667    18.900000   1.333333    0.333333       0.333333  
7     3.666667    17.733333   0.666667    0.000000       0.000000  
9     4.333333    18.233333   0.666667    0.000000       0.000000  
11    6.666667    18.466667   1.000000    0.000000       0.000000  
12    5.666667    19.366667   1.000000    0.000000       0.333333  
..         ...          ...        ...         ...            ...  
42    7.000000    15.366667   0.333333    0.333333       0.333333  
44    7.333333    16.000000   0.333333    0.000000       0.000000  
46    6.666667    16.133333   0.333333    0.000000       0.000000  
49    4.666667    16.700000   0.333333    0.000000       0.000000  
50    5.000000    17.200000   0.333333    0.000000       0.000000  

[68 rows x 40 columns]

In [76]:
pertandingan_rolling = pertandingan.groupby("team").apply(lambda x: rolling_averages(x, cols, new_cols))

In [77]:
pertandingan_rolling

date   time            comp         round  \
team                                                                         
Arsenal                 6  2020-10-04  14:00  Premier League   Matchweek 4   
                        7  2020-10-17  17:30  Premier League   Matchweek 5   
                        9  2020-10-25  19:15  Premier League   Matchweek 6   
                        11 2020-11-01  16:30  Premier League   Matchweek 7   
                        13 2020-11-08  19:15  Premier League   Matchweek 8   
...                               ...    ...             ...           ...   
Wolverhampton Wanderers 32 2022-03-13  14:00  Premier League  Matchweek 29   
                        33 2022-03-18  20:00  Premier League  Matchweek 30   
                        34 2022-04-02  15:00  Premier League  Matchweek 31   
                        35 2022-04-08  20:00  Premier League  Matchweek 32   
                        36 2022-04-24  14:00  Premier League  Matchweek 34   

                            day venue result   gf   ga         opponent  ...  \
team                                                                     ...   
Arsenal                 6   Sun  Home      W  2.0  1.0    Sheffield Utd  ...   
                        7   Sat  Away      L  0.0  1.0  Manchester City  ...   
                        9   Sun  Home      L  0.0  1.0   Leicester City  ...   
                        11  Sun  Away      W  1.0  0.0   Manchester Utd  ...   
                        13  Sun  Home      L  0.0  3.0      Aston Villa  ...   
...                         ...   ...    ...  ...  ...              ...  ...   
Wolverhampton Wanderers 32  Sun  Away      W  1.0  0.0          Everton  ...   
                        33  Fri  Home      L  2.0  3.0     Leeds United  ...   
                        34  Sat  Home      W  2.0  1.0      Aston Villa  ...   
                        35  Fri  Away      L  0.0  1.0    Newcastle Utd  ...   
                        36  Sun  Away      L  0.0  1.0          Burnley  ...   

                            hari  target  gf_rolling  ga_rolling sh_rolling  \
team                                                                          
Arsenal                 6      6       1    2.000000    1.333333   7.666667   
                        7      5       0    1.666667    1.666667   5.333333   
                        9      6       0    1.000000    1.666667   7.000000   
                        11     6       1    0.666667    1.000000   9.666667   
                        13     6       0    0.333333    0.666667   9.666667   
...                          ...     ...         ...         ...        ...   
Wolverhampton Wanderers 32     6       1    1.333333    1.000000  12.333333   
                        33     4       0    1.666667    0.666667  12.333333   
                        34     5       1    2.333333    1.000000  13.000000   
                        35     4       0    1.666667    1.333333  13.000000   
                        36     6       0    1.333333    1.666667  10.000000   

                           sot_rolling dist_rolling fk_rolling  pk_rolling  \
team                                                                         
Arsenal                 6     3.666667    14.733333   0.666667    0.000000   
                        7     3.666667    15.766667   0.000000    0.000000   
                        9     3.666667    16.733333   0.666667    0.000000   
                        11    4.000000    16.033333   1.000000    0.000000   
                        13    2.666667    18.033333   1.000000    0.333333   
...                                ...          ...        ...         ...   
Wolverhampton Wanderers 32    3.666667    19.300000   0.000000    0.000000   
                        33    4.333333    19.600000   0.000000    0.000000   
                        34    5.333333    19.833333   0.000000    0.000000   
                        35    5.000000    18.533333   0.000000    0.000000   
             

In [78]:
pertandingan_rolling=pertandingan_rolling.droplevel("team")
pertandingan_rolling

date   time            comp         round  day venue result   gf  \
6  2020-10-04  14:00  Premier League   Matchweek 4  Sun  Home      W  2.0   
7  2020-10-17  17:30  Premier League   Matchweek 5  Sat  Away      L  0.0   
9  2020-10-25  19:15  Premier League   Matchweek 6  Sun  Home      L  0.0   
11 2020-11-01  16:30  Premier League   Matchweek 7  Sun  Away      W  1.0   
13 2020-11-08  19:15  Premier League   Matchweek 8  Sun  Home      L  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
32 2022-03-13  14:00  Premier League  Matchweek 29  Sun  Away      W  1.0   
33 2022-03-18  20:00  Premier League  Matchweek 30  Fri  Home      L  2.0   
34 2022-04-02  15:00  Premier League  Matchweek 31  Sat  Home      W  2.0   
35 2022-04-08  20:00  Premier League  Matchweek 32  Fri  Away      L  0.0   
36 2022-04-24  14:00  Premier League  Matchweek 34  Sun  Away      L  0.0   

     ga         opponent  ...  hari  target  gf_rolling  ga_rolling  \
6   1.0    Sheffield Utd  ...     6       1    2.000000    1.333333   
7   1.0  Manchester City  ...     5       0    1.666667    1.666667   
9   1.0   Leicester City  ...     6       0    1.000000    1.666667   
11  0.0   Manchester Utd  ...     6       1    0.666667    1.000000   
13  3.0      Aston Villa  ...     6       0    0.333333    0.666667   
..  ...              ...  ...   ...     ...         ...         ...   
32  0.0          Everton  ...     6       1    1.333333    1.000000   
33  3.0     Leeds United  ...     4       0    1.666667    0.666667   
34  1.0      Aston Villa  ...     5       1    2.333333    1.000000   
35  1.0    Newcastle Utd  ...     4       0    1.666667    1.333333   
36  1.0          Burnley  ...     6       0    1.333333    1.666667   

   sh_rolling sot_rolling dist_rolling fk_rolling  pk_rolling  pkatt_rolling  
6    7.666667    3.666667    14.733333   0.666667    0.000000       0.000000  
7    5.333333    3.666667    15.766667   0.000000    0.000000       0.000000  
9    7.000000    3.666667    16.733333   0.666667    0.000000       0.000000  
11   9.666667    4.000000    16.033333   1.000000    0.000000       0.000000  
13   9.666667    2.666667    18.033333   1.000000    0.333333       0.333333  
..        ...         ...          ...        ...         ...            ...  
32  12.333333    3.666667    19.300000   0.000000    0.000000       0.000000  
33  12.333333    4.333333    19.600000   0.000000    0.000000       0.000000  
34  13.000000    5.333333    19.833333   0.000000    0.000000       0.000000  
35  13.000000    5.000000    18.533333   0.000000    0.000000       0.000000  
36  10.000000    4.666667    17.633333   0.000000    0.000000       0.000000  

[1317 rows x 40 columns]

Menghapus Kolom tim agar lebih mudah dalam menghitung algoritma(Kalau tidak harus pakai 2 value)

In [79]:
pertandingan_rolling.index=range(pertandingan_rolling.shape[0])
pertandingan_rolling

date   time            comp         round  day venue result   gf  \
0    2020-10-04  14:00  Premier League   Matchweek 4  Sun  Home      W  2.0   
1    2020-10-17  17:30  Premier League   Matchweek 5  Sat  Away      L  0.0   
2    2020-10-25  19:15  Premier League   Matchweek 6  Sun  Home      L  0.0   
3    2020-11-01  16:30  Premier League   Matchweek 7  Sun  Away      W  1.0   
4    2020-11-08  19:15  Premier League   Matchweek 8  Sun  Home      L  0.0   
...         ...    ...             ...           ...  ...   ...    ...  ...   
1312 2022-03-13  14:00  Premier League  Matchweek 29  Sun  Away      W  1.0   
1313 2022-03-18  20:00  Premier League  Matchweek 30  Fri  Home      L  2.0   
1314 2022-04-02  15:00  Premier League  Matchweek 31  Sat  Home      W  2.0   
1315 2022-04-08  20:00  Premier League  Matchweek 32  Fri  Away      L  0.0   
1316 2022-04-24  14:00  Premier League  Matchweek 34  Sun  Away      L  0.0   

       ga         opponent  ...  hari  target  gf_rolling  ga_rolling  \
0     1.0    Sheffield Utd  ...     6       1    2.000000    1.333333   
1     1.0  Manchester City  ...     5       0    1.666667    1.666667   
2     1.0   Leicester City  ...     6       0    1.000000    1.666667   
3     0.0   Manchester Utd  ...     6       1    0.666667    1.000000   
4     3.0      Aston Villa  ...     6       0    0.333333    0.666667   
...   ...              ...  ...   ...     ...         ...         ...   
1312  0.0          Everton  ...     6       1    1.333333    1.000000   
1313  3.0     Leeds United  ...     4       0    1.666667    0.666667   
1314  1.0      Aston Villa  ...     5       1    2.333333    1.000000   
1315  1.0    Newcastle Utd  ...     4       0    1.666667    1.333333   
1316  1.0          Burnley  ...     6       0    1.333333    1.666667   

     sh_rolling sot_rolling dist_rolling fk_rolling  pk_rolling  pkatt_rolling  
0      7.666667    3.666667    14.733333   0.666667    0.000000       0.000000  
1      5.333333    3.666667    15.766667   0.000000    0.000000       0.000000  
2      7.000000    3.666667    16.733333   0.666667    0.000000       0.000000  
3      9.666667    4.000000    16.033333   1.000000    0.000000       0.000000  
4      9.666667    2.666667    18.033333   1.000000    0.333333       0.333333  
...         ...         ...          ...        ...         ...            ...  
1312  12.333333    3.666667    19.300000   0.000000    0.000000       0.000000  
1313  12.333333    4.333333    19.600000   0.000000    0.000000       0.000000  
1314  13.000000    5.333333    19.833333   0.000000    0.000000       0.000000  
1315  13.000000    5.000000    18.533333   0.000000    0.000000       0.000000  
1316  10.000000    4.666667    17.633333   0.000000    0.000000       0.000000  

[1317 rows x 40 columns]

# Melatih ulang Kode

In [82]:
def buat_prediksi(data, acuan):
    train=data[data["date"] < '2022-01-01']
    test=data[data["date"] > '2022-01-01']
    rf.fit(train[acuan], train["target"])
    prediksi = rf.predict(test[acuan])
    combined=pd.DataFrame(dict(actual=test["target"], predicted=prediksi), index=test.index)
    precision = precision_score(test["target"], prediksi)
    return combined, precision

# Map

In [83]:
combined, precision = buat_prediksi(pertandingan_rolling, acuan+new_cols)

Melihat Akurasi yang baru

In [86]:
precision

0.625

In [87]:
combined

actual  predicted
55         0          0
56         1          0
57         1          0
58         1          1
59         1          1
...      ...        ...
1312       1          0
1313       0          0
1314       1          0
1315       0          0
1316       0          0

[276 rows x 2 columns]

Disini kita belum bisa melihat ini adalah data dari tim mana
untuk itu perlu dilakukan perubahan

In [89]:
combined = combined.merge(pertandingan_rolling[["date", "team", "opponent", "result"]], left_index=True, right_index=True)

In [90]:
combined

actual  predicted       date                     team       opponent  \
55         0          0 2022-01-23                  Arsenal        Burnley   
56         1          0 2022-02-10                  Arsenal         Wolves   
57         1          0 2022-02-19                  Arsenal      Brentford   
58         1          1 2022-02-24                  Arsenal         Wolves   
59         1          1 2022-03-06                  Arsenal        Watford   
...      ...        ...        ...                      ...            ...   
1312       1          0 2022-03-13  Wolverhampton Wanderers        Everton   
1313       0          0 2022-03-18  Wolverhampton Wanderers   Leeds United   
1314       1          0 2022-04-02  Wolverhampton Wanderers    Aston Villa   
1315       0          0 2022-04-08  Wolverhampton Wanderers  Newcastle Utd   
1316       0          0 2022-04-24  Wolverhampton Wanderers        Burnley   

     result  
55        D  
56        W  
57        W  
58        W  
59        W  
...     ...  
1312      W  
1313      L  
1314      W  
1315      L  
1316      L  

[276 rows x 6 columns]

Dalam data ini terdapat 2 tim dimana ada yang menjadi team dan ada yang menjadi opponent untuk itu kita akan melakukan merge agar data lebih akurat.

Akan tetapi dalam data tersebut nama team dan oppenent terkadang tidak sama untuk itu perlu dilakukan nomralize

In [91]:
class MissingDict(dict):
    __missing__ = lambda self, key : key

map_values = {"Brighton and Hove Albion": "Brighton", 
              "Manchester United": "Manchester Utd", 
              "Newcastle United": "Newcastle Utd", 
              "Tottenham Hotspur": "Tottenham", 
              "West Ham United": "West Ham", 
              "Wolverhampton Wanderers": "Wolves"
}
mapping=MissingDict(**map_values)    


In [93]:
combined["tim_baru"]= combined["team"].map(mapping)

In [94]:
combined

actual  predicted       date                     team       opponent  \
55         0          0 2022-01-23                  Arsenal        Burnley   
56         1          0 2022-02-10                  Arsenal         Wolves   
57         1          0 2022-02-19                  Arsenal      Brentford   
58         1          1 2022-02-24                  Arsenal         Wolves   
59         1          1 2022-03-06                  Arsenal        Watford   
...      ...        ...        ...                      ...            ...   
1312       1          0 2022-03-13  Wolverhampton Wanderers        Everton   
1313       0          0 2022-03-18  Wolverhampton Wanderers   Leeds United   
1314       1          0 2022-04-02  Wolverhampton Wanderers    Aston Villa   
1315       0          0 2022-04-08  Wolverhampton Wanderers  Newcastle Utd   
1316       0          0 2022-04-24  Wolverhampton Wanderers        Burnley   

     result tim_baru  
55        D  Arsenal  
56        W  Arsenal  
57        W  Arsenal  
58        W  Arsenal  
59        W  Arsenal  
...     ...      ...  
1312      W   Wolves  
1313      L   Wolves  
1314      W   Wolves  
1315      L   Wolves  
1316      L   Wolves  

[276 rows x 7 columns]

In [95]:
merged=combined.merge(combined, left_on=["date", "tim_baru"], right_on=["date", "opponent"])
merged

actual_x  predicted_x       date                   team_x     opponent_x  \
0           0            0 2022-01-23                  Arsenal        Burnley   
1           1            0 2022-02-10                  Arsenal         Wolves   
2           1            0 2022-02-19                  Arsenal      Brentford   
3           1            1 2022-02-24                  Arsenal         Wolves   
4           1            1 2022-03-06                  Arsenal        Watford   
..        ...          ...        ...                      ...            ...   
257         1            0 2022-03-13  Wolverhampton Wanderers        Everton   
258         0            0 2022-03-18  Wolverhampton Wanderers   Leeds United   
259         1            0 2022-04-02  Wolverhampton Wanderers    Aston Villa   
260         0            0 2022-04-08  Wolverhampton Wanderers  Newcastle Utd   
261         0            0 2022-04-24  Wolverhampton Wanderers        Burnley   

    result_x tim_baru_x  actual_y  predicted_y                   team_y  \
0          D    Arsenal         0            0                  Burnley   
1          W    Arsenal         0            0  Wolverhampton Wanderers   
2          W    Arsenal         0            0                Brentford   
3          W    Arsenal         0            0  Wolverhampton Wanderers   
4          W    Arsenal         0            0                  Watford   
..       ...        ...       ...          ...                      ...   
257        W     Wolves         0            0                  Everton   
258        L     Wolves         1            0             Leeds United   
259        W     Wolves         0            0              Aston Villa   
260        L     Wolves         1            0         Newcastle United   
261        L     Wolves         1            0                  Burnley   

    opponent_y result_y     tim_baru_y  
0      Arsenal        D        Burnley  
1      Arsenal        L         Wolves  
2      Arsenal        L      Brentford  
3      Arsenal        L         Wolves  
4      Arsenal        L        Watford  
..         ...      ...            ...  
257     Wolves        L        Everton  
258     Wolves        W   Leeds United  
259     Wolves        L    Aston Villa  
260     Wolves        W  Newcastle Utd  
261     Wolves        W        Burnley  

[262 rows x 13 columns]

In [101]:
print ("Akurasi = ", 27/40)
merged[(merged["predicted_x"] == 1) & (merged["predicted_y"] ==0)]["actual_x"].value_counts()


Akurasi =  0.675


1    27
0    13
Name: actual_x, dtype: int64

In [149]:
hasil_prediksi=merged.rename(columns = { 
    "tim_baru_x":"Tim",
    "opponent_x":"Lawan",
    "result_x":"Prediksi"
})
hasil_prediksi=hasil_prediksi[["date","Tim", "Lawan", "Prediksi"]]
hasil_prediksi

date      Tim          Lawan Prediksi
0   2022-01-23  Arsenal        Burnley        D
1   2022-02-10  Arsenal         Wolves        W
2   2022-02-19  Arsenal      Brentford        W
3   2022-02-24  Arsenal         Wolves        W
4   2022-03-06  Arsenal        Watford        W
..         ...      ...            ...      ...
257 2022-03-13   Wolves        Everton        W
258 2022-03-18   Wolves   Leeds United        L
259 2022-04-02   Wolves    Aston Villa        W
260 2022-04-08   Wolves  Newcastle Utd        L
261 2022-04-24   Wolves        Burnley        L

[262 rows x 4 columns]

In [150]:
hasil_prediksi.to_csv('prediksi.csv',sep=',')

In [151]:
dprediksi = pd.read_csv("prediksi.csv", index_col=0)
dprediksi

date      Tim          Lawan Prediksi
0    2022-01-23  Arsenal        Burnley        D
1    2022-02-10  Arsenal         Wolves        W
2    2022-02-19  Arsenal      Brentford        W
3    2022-02-24  Arsenal         Wolves        W
4    2022-03-06  Arsenal        Watford        W
..          ...      ...            ...      ...
257  2022-03-13   Wolves        Everton        W
258  2022-03-18   Wolves   Leeds United        L
259  2022-04-02   Wolves    Aston Villa        W
260  2022-04-08   Wolves  Newcastle Utd        L
261  2022-04-24   Wolves        Burnley        L

[262 rows x 4 columns]

In [155]:
dprediksi[(dprediksi["Tim"] == "Manchester City") & (dprediksi["Lawan"]=="Chelsea")]

date              Tim    Lawan Prediksi
136  2022-01-15  Manchester City  Chelsea        W